mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp

```diff
@@ -0,0 +1,501 @@
+      auto insertRHS =
+          rewriter.create<vector::ScalableInsertOp>(loc, rhs, undef, 0);
+      return rewriter.create<arm_sve::Zip1IntrOp>(loc, type, insertLHS,
+                                                  insertRHS);
+    };
+
+    auto extOp = op.getLhs().getDefiningOp();
+    VectorType extSourceVectorType =
+        cast<VectorType>(extOp->getOperand(0).getType());
+    VectorType widenedVectorType =
+        VectorType::Builder(extSourceVectorType)
+            .setDim(0, extSourceVectorType.getShape()[0] * 2);
+    auto lhs = packInputs(widenedVectorType,
+                          op1.getLhs().getDefiningOp()->getOperand(0),
+                          op2.getLhs().getDefiningOp()->getOperand(0));
+    auto rhs = packInputs(widenedVectorType,
+                          op1.getRhs().getDefiningOp()->getOperand(0),
+                          op2.getRhs().getDefiningOp()->getOperand(0));
+
+    Value lhsMask, rhsMask;
+    if (op1.getLhsMask() || op2.getLhsMask()) {
+      if (!(op1.getLhsMask() && op2.getLhsMask()))
+        return rewriter.notifyMatchFailure(
+            op, "unsupported masking, either both outerproducts are masked "
+                "or neither");
+
+      VectorType maskType = VectorType::Builder(widenedVectorType)
+                                .setElementType(rewriter.getI1Type());
+      lhsMask = packInputs(maskType, op1.getLhsMask(), op2.getLhsMask());
+      rhsMask = packInputs(maskType, op1.getRhsMask(), op2.getRhsMask());
+    }
+
+    arm_sme::CombiningKind kind = op.getKind();
+    assert((kind == arm_sme::CombiningKind::Add ||
+            kind == arm_sme::CombiningKind::Sub) &&
+           "unhandled arm_sme::CombiningKind!");
+
+    if (isa<arith::ExtFOp>(extOp)) {
+      if (kind == arm_sme::CombiningKind::Add)
+        rewriter.replaceOpWithNewOp<arm_sme::FMopaWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        rewriter.replaceOpWithNewOp<arm_sme::FMopsWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+    } else if (isa<arith::ExtSIOp>(extOp)) {
+      if (kind == arm_sme::CombiningKind::Add)
+        rewriter.replaceOpWithNewOp<arm_sme::SMopaWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        rewriter.replaceOpWithNewOp<arm_sme::SMopsWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+    } else if (isa<arith::ExtUIOp>(extOp)) {
+      if (kind == arm_sme::CombiningKind::Add)
+        rewriter.replaceOpWithNewOp<arm_sme::UMopaWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        rewriter.replaceOpWithNewOp<arm_sme::UMopsWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+    } else
+      llvm_unreachable("unexpected extend op!");
+
+    op1.erase();
+
+    return success();
```

Comment:
IMHO, `else` obscures the flow. This might be a matter of style, so feel free to ignore.

```suggestion
if (kind == arm_sme::CombiningKind::Add) {
  if (isa<arith::ExtFOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::FMopaWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtSIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::SMopaWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtUIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::UMopaWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  llvm_unreachable("unexpected extend op!");
}

if (kind == arm_sme::CombiningKind::Sub) {
  if (isa<arith::ExtFOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::FMopsWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtSIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::SMopsWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtUIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::UMopsWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  llvm_unreachable("unexpected extend op!");
}

op1.erase();
return success();
```

Reply:
Or an approach I like with TypeSwitch:
```c++
  if (kind == arm_sme::CombiningKind::Add) {
    TypeSwitch<Operation *>(extOp)
        .Case<cf::ExtFOp>([&](auto) {
          rewriter.replaceOpWithNewOp<arm_sme::FMopaWide2WayOp>(
              op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask,
              op1.getAcc());
        })
        .Case<cf::ExtSIOp>([&](auto) {
          rewriter.replaceOpWithNewOp<arm_sme::SMopaWide2WayOp>(
              op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask,
              op1.getAcc());
        })
        .Case<ExtUIOp>([&](auto) {
          rewriter.replaceOpWithNewOp<arm_sme::UMopaWide2WayOp>(
              op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask,
              op1.getAcc());
        })
        .Default([&](auto) { llvm_unreachable("unexpected extend op!"); });
  }

  /// ... same again for Sub ...
```